pkg/ring/replication_set_test.go

```diff
@@ -33,3 +38,226 @@ func TestReplicationSet_GetAddresses(t *testing.T) {
+		f                   func(context.Context, *IngesterDesc) (interface{}, error)
+		delay               time.Duration
+		cancelContextDelay  time.Duration
+		want                []interface{}
+		expectedError       error
+	}{
+		{
+			name: "max errors = 0, no errors no delay",
+			instances: []IngesterDesc{
+				{},
+			},
+			f: func(c context.Context, id *IngesterDesc) (interface{}, error) {
+				return 1, nil
+			},
+			want: []interface{}{1},
+		},
+		{
+			name:      "max errors = 0, should fail on 1 error out of 1 instance",
+			instances: []IngesterDesc{{}},
+			f: func(c context.Context, id *IngesterDesc) (interface{}, error) {
+				return nil, errFailure
+			},
+			want:          nil,
+			expectedError: errFailure,
+		},
+		{
+			name:          "max errors = 0, should fail on 1 error out of 3 instances (last call fails)",
+			instances:     []IngesterDesc{{}, {}, {}},
+			f:             failingFunctionAfter(2, 10*time.Millisecond),
+			want:          nil,
+			expectedError: errFailure,
+		},
+		{
+			name:          "max errors = 1, should fail on 3 errors out of 5 instances (last calls fail)",
+			instances:     []IngesterDesc{{}, {}, {}, {}, {}},
+			maxErrors:     1,
+			f:             failingFunctionAfter(2, 10*time.Millisecond),
+			delay:         100 * time.Millisecond,
+			want:          nil,
+			expectedError: errFailure,
+		},
+		{
+			name:      "max errors = 1, should handle context canceled",
+			instances: []IngesterDesc{{}, {}, {}},
+			maxErrors: 1,
+			f: func(c context.Context, id *IngesterDesc) (interface{}, error) {
+				time.Sleep(300 * time.Millisecond)
+				return 1, nil
+			},
+			cancelContextDelay: 100 * time.Millisecond,
+			want:               nil,
+			expectedError:      context.Canceled,
+		},
+		{
+			name: "max errors = 0, should succeed on all successful instances",
+			instances: []IngesterDesc{{
+				Zone: "zone1",
+			}, {
+				Zone: "zone2",
+			}, {
+				Zone: "zone3",
+			}},
+			f: func(c context.Context, id *IngesterDesc) (interface{}, error) {
+				return 1, nil
+			},
+			want: []interface{}{1, 1, 1},
+		},
+		{
+			name: "max unavailable zones = 1, should succeed on instances failing in 1 out of 3 zones (3 instances)",
+			instances: []IngesterDesc{{
+				Zone: "zone1",
+			}, {
+				Zone: "zone2",
+			}, {
+				Zone: "zone3",
+			}},
+			f:                   failingFunctionOnZones("zone1"),
+			maxUnavailableZones: 1,
+			want:                []interface{}{1, 1},
+		},
+		{
+			name: "max unavailable zones = 1, should fail on instances failing in 2 out of 3 zones (3 instances)",
+			instances: []IngesterDesc{{
+				Zone: "zone1",
+			}, {
+				Zone: "zone2",
+			}, {
+				Zone: "zone3",
+			}},
+			f:                   failingFunctionOnZones("zone1", "zone2"),
+			maxUnavailableZones: 1,
+			expectedError:       errZoneFailure,
+		},
+		{
+			name: "max unavailable zones = 1, should succeed on instances failing in 1 out of 3 zones (6 instances)",
+			instances: []IngesterDesc{{
```

Comment:
Can we reformat these long vertical blocks to be more compact and readable? (In entire test)

Reply:
Any suggestion how to achieve that? I could add a function to create the list of instances which could perhaps also return the expectedAddresses. Not sure that would be much more readable though.

Reply:
I didn't mean anything fancy, just removing extra newlines would go a long way here. Eg.:

```
			instances:           []IngesterDesc{{Zone: "zone1"}, {Zone: "zone1"}, {Zone: "zone2"}, {Zone: "zone2"}, {Zone: "zone3"}, {Zone: "zone3"}},
```

instead of 

```
			instances: []IngesterDesc{{
				Zone: "zone1",
			}, {
				Zone: "zone1",
			}, {
				Zone: "zone2",
			}, {
				Zone: "zone2",
			}, {
				Zone: "zone3",
			}, {
				Zone: "zone3",
			}}
```

Reply:
My mistake, I was looking in the wrong place. The newlines have been removed.